packages/react-dom-bindings/src/client/ReactDOMFloatClient.js

```diff
@@ -1390,40 +802,220 @@ function insertStyleInstance(
-    if (parent) {
-      parent.insertBefore(instance, parent.firstChild);
-    } else {
-      throw new Error(
-        'While attempting to insert a Resource, React expected the Document to contain' +
-          ' a head element but it was not found.',
-      );
-    }
+      root.nodeType === DOCUMENT_NODE
+        ? ((((root: any): Document).head: any): Element)
+        : ((root: any): ShadowRoot);
+    parent.insertBefore(instance, parent.firstChild);
   }
 }
 
-function insertResourceInstanceBefore(
-  ownerDocument: Document,
-  instance: Instance,
-  before: ?Node,
+function adoptPreloadPropsForStylesheet(
+  stylesheetProps: StylesheetProps,
+  preloadProps: PreloadProps,
 ): void {
-  if (__DEV__) {
-    if (instance.tagName === 'LINK' && (instance: any).rel === 'stylesheet') {
-      console.error(
-        'insertResourceInstanceBefore was called with a stylesheet. Stylesheets must be' +
-          ' inserted with insertStyleInstance instead. This is a bug in React.',
-      );
-    }
+  if (stylesheetProps.crossOrigin == null)
+    stylesheetProps.crossOrigin = preloadProps.crossOrigin;
+  if (stylesheetProps.referrerPolicy == null)
+    stylesheetProps.referrerPolicy = preloadProps.referrerPolicy;
+  if (stylesheetProps.title == null) stylesheetProps.title = preloadProps.title;
+}
+
+function adoptPreloadPropsForScript(
+  scriptProps: ScriptProps,
+  preloadProps: PreloadProps,
+): void {
+  if (scriptProps.crossOrigin == null)
+    scriptProps.crossOrigin = preloadProps.crossOrigin;
+  if (scriptProps.referrerPolicy == null)
+    scriptProps.referrerPolicy = preloadProps.referrerPolicy;
+  if (scriptProps.integrity == null)
+    scriptProps.referrerPolicy = preloadProps.integrity;
+}
+
+// --------------------------------------
+//      Hoistable Element Reconciliation
+// --------------------------------------
+
+export function hydrateHoistable(
+  hoistableRoot: HoistableRoot,
+  type: HoistableTagType,
+  props: any,
+  internalInstanceHandle: Object,
+): Instance {
+  const ownerDocument = getDocumentFromRoot(hoistableRoot);
+  const nodes = ownerDocument.querySelectorAll(type);
+
+  const children = props.children;
+  let child, childString;
+  if (Array.isArray(children)) {
+    child = children.length === 1 ? children[0] : null;
+  } else {
+    child = children;
   }
-  const parent = (before && before.parentNode) || ownerDocument.head;
-  if (parent) {
-    parent.insertBefore(instance, before);
+  if (
+    typeof child !== 'function' &&
+    typeof child !== 'symbol' &&
+    child !== null &&
+    child !== undefined
+  ) {
+    if (__DEV__) {
+      checkPropStringCoercion(child, 'children');
+    }
+    childString = '' + (child: any);
   } else {
-    throw new Error(
-      'While attempting to insert a Resource, React expected the Document to contain' +
-        ' a head element but it was not found.',
-    );
+    childString = '';
+  }
+  nodeLoop: for (let i = 0; i < nodes.length; i++) {
```

Comment:
I'm not loving the computational complexity of N * N queries for number of nodes. I wonder if we need to do something like caching in a Map.

Reply:
yeah same, i'll think about more efficient implementations for a follow up